File: java/Magic_Triplets.java

/* 

Given an array of size n, a triplet (a[i], a[j], a[k]) is called a Magic Triplet if a[i] < a[j] < a[k] and i < j < k.  
Count the number of magic triplets in a given array.
 

Example 1:

Input: arr = [3, 2, 1]
Output: 0
Explanation: There is no magic triplet.

Example 2:

Input: arr = [1, 2, 3, 4]
Output: 4
Explanation: Fours magic triplets are 
(1, 2, 3), (1, 2, 4), (1, 3, 4) and 
(2, 3, 4).
 

Your Task:
You don't need to read or print anything. Your task is to complete the function countTriplets() which takes the array nums[] as input parameter and 
returns the number of magic triplets in the array.

 
Expected Time Complexity: O(N2) 
Expected Space Complexity: O(1)
*/


class Solution{
    public int countTriplets(int[] nums){
        // code here
        int ans=0;
        int n=nums.length;
        for(int i=0;i<n;i++){
            int left=0,right=0;
	        // no. smaller than a[i] would be in the left 
	        for(int j=i-1;j>=0;j--){
	            if(nums[i]>nums[j]) left++;
	        }
	         // no. larger than a[i] would be int the right 
	         for(int k=i+1;k<n;k++)
	         {
	             if(nums[i]<nums[k]) right++;
	         }
	         ans+=(left*right);
        }
        return ans;
    }
}

File: Arrays/MagicTriplets.cpp
#include <iostream>
#include <vector>
using namespace std;

int countMagicTriplets(const vector<int>& arr) {
    int n = arr.size();
    int count = 0;

    // Traverse each element as the middle element
    for (int j = 1; j < n - 1; ++j) {
        int left_count = 0;
        int right_count = 0;

        // Count smaller elements on the left of arr[j]
        for (int i = 0; i < j; ++i) {
            if (arr[i] < arr[j]) {
                ++left_count;
            }
        }

        // Count larger elements on the right of arr[j]
        for (int k = j + 1; k < n; ++k) {
            if (arr[k] > arr[j]) {
                ++right_count;
            }
        }

        // Add the number of triplets with arr[j] as the middle element
        count += left_count * right_count;
    }

    return count;
}

int main() {
    vector<int> arr = {3, 4, 7, 7, 11, 3, 9, 6, 8};
    cout << "Number of magic triplets: " << countMagicTriplets(arr) << endl;
    return 0;
}
